### DISCUSSION TITLE
Number of tiles, or points budget

### DISCUSSION THREAD

**Discussion by ltphy** [visgl/loaders.gl#1527] [Q&A] [answered]
I have tested potree library, which is also a 3d visualization library. There is an option called point budget in that library, which will only render number of nodes depending on the allowed numbers of points and all of the leave nodes will be removed accordingly.

![point_budget](https://user-images.githubusercontent.com/32784614/122669880-72e30d80-d1e9-11eb-9ef7-e0c18bb9cfd3.png)

Is there any function in tileset-3d that can execute similarly?
If the function is currently not available, what should I implement in the code to have that similar feature?

**Comment by ibgreen** [comment, marked as answer]
This seems to be a good feature to handle improperly tiled datasets (the whole point of tiling datasets is to make sure that only a reasonable amount of geometry is selected for rendering at any time).

I think this could be done fairly independently as a post processing pass after the current traversal algorithm (Tileset3D.update) where we would walk the selected tiles top down and remove child tiles when budget is exceeded.

The main wrinkle is that we support multiviewport traversal, so the post processing should be done on each viewport.

If we implement in loaders, we should handle both the point cloud and mesh cases - it should not be much of a difference.

I suppose the post processing could go in here https://github.com/visgl/loaders.gl/blob/master/modules/tiles/src/tileset/tileset-3d.ts#L435

@belom88 thougths?

**Reply by ltphy** [reply]
Hi @ibgreen. Thank you for your suggesstions.
I have tried to load all the selected tiles by traversing BFS from the root until meet two qualification: a selected tile in the current frame and the accumulative number of point cloud is still under the point budget allowance.
However, there will be cases that center selected tiles are not shown on the screen. (they are not selected

My second try is to sort all the selected tiles of the current frame in ascending order by the camera distance, and take the selected tiles until the total number of point cloud exceeds the point budget. This method works better for me.
  
Your thoughts?

**Reply by belom88** [reply]
https://github.com/visgl/loaders.gl/blob/a21a0d49c3b6869591016e1c289c1a9fb7f0b2b6/modules/tiles/src/tileset/tileset-3d.ts#L486
this place is better. This loop joins tiles, selected for all viewports. So we could do some post-process filtering there.
I tried doing this filter during traversal. It is not good idea because refinement and frustum culling algorithms as quite difficult.

**Reply by ibgreen** [reply]
Perhaps we could offer a Tileset3D callback to let the user implement/experiment with tile filtering. Something like
`Tileset3D.options.onTraversalComplete(...)`

**Reply by belom88** [reply]
I agree. Let me do this thing.

**Reply by ltphy** [reply]
One more thing, after filtering out the selected tiles, I have to delete the remaining tiles to prevent the cacheMap to access those that are end of their lifecycle as well.

**Reply by ibgreen** [reply]
Assuming this is the cache in the deck.gl layer: if the Tile3DLayer doesn't see the tile (assuming filtering in `onTraversalComplete` prevents the tiles from being returned from `tileset.update` etc), they should not end up in the layer's cache.

**Reply by belom88** [reply]
https://github.com/visgl/loaders.gl/pull/1596
